File: CS/ChartsDemo.Wpf/ViewModels/MathematicalFunctionsViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Windows.Media.Imaging;
using System.Windows.Media.Media3D;
using DevExpress.Xpf.DemoBase;
using DevExpress.Xpf.Editors;
using System.Linq;
using DevExpress.Mvvm;

namespace ChartsDemo {
    public class MathematicalFunctionsViewModel {
        const double UnitFactor = 15;

        public List<Function3DItemData> FunctionItemsData { get; private set; }
        public virtual Function3DItemData Function { get; set; }

        public MathematicalFunctionsViewModel() {
            FunctionItemsData = CreateFunctionItemsData();
            Function = FunctionItemsData[3];
        }

        #region Static

        static List<Function3DItemData> CreateFunctionItemsData() {
            var funcs = new Func<double, double, Point3D>[] 
                { CalculateFirstValue, CalculateSecondValue, CalculateThirdValue, CalculateFourthValue, CalculateFifthValue };
            var images = Enumerable.Range(1, 5).Select(x => Chart3DUtils.CreateFunctionImage(x));
            return funcs.Zip(images, (func, image) => new Function3DItemData { Image = image, Points = CreatePoints(func) }).ToList();
        }

        static double Sinc(double x) {
            return x != 0 ? Math.Sin(x) : 1;
        }
        static Point3D CalculateFirstValue(double x, double y) {
            x *= UnitFactor;
            y *= UnitFactor;
            double value = Sinc(Math.Sin(Math.Pow(Math.Pow(x, 6) + Math.Pow(y, 6), 1.0d / 6.0d))) * 5;
            return new Point3D(x, y, value);
        }
        static Point3D CalculateSecondValue(double x, double y) {
            x *= UnitFactor;
            y *= UnitFactor;
            double square = Math.Sqrt(x * x + y * y);
            double value = square * Sinc(square) * 0.2;
            return new Point3D(x, y, value);
        }
        static Point3D CalculateThirdValue(double x, double y) {
            x *= UnitFactor / 2;
            y *= UnitFactor / 2;
            double value = Math.Sin(x) * Math.Cos(y) * 2;
            return new Point3D(x, y, value);
        }
        static Point3D CalculateFourthValue(double x, double y) {
            double theta = Math.Atan2(y, x);
            double r = x * x + y * y;
            double z = Math.Exp(-r) * Math.Sin(2 * Math.PI * Math.Sqrt(r)) * Math.Cos(3 * theta);
            return new Point3D(x, y, z);
        }
        static Point3D CalculateFifthValue(double x, double y) {
            x *= 3;
            y *= 3;
            double z = Math.Sin(x * y);
            return new Point3D(x, y, z);
        }
        static List<Point3D> CreatePoints(Func<double, double, Point3D> valueCalculator) {
            List<Point3D> points = new List<Point3D>();
            for(double x = -1; x < 1; x += 0.017)
                for(double y = -1; y < 1; y += 0.017)
                    points.Add(valueCalculator(x, y));
            return points;
        }

        #endregion
    }
    public class Function3DItemData : BindableBase {
        List<Point3D> points;
        BitmapImage image;

        public List<Point3D> Points {
            get { return points; }
            set { SetProperty(ref points, value, () => Points); }
        }
        public BitmapImage Image {
            get { return image; }
            set { SetProperty(ref image, value, () => Image); }
        }
    }
}

File: VB/ChartsDemo.Wpf/ViewModels/MathematicalFunctionsViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Windows.Media.Imaging
Imports System.Windows.Media.Media3D
Imports DevExpress.Xpf.DemoBase
Imports DevExpress.Xpf.Editors
Imports System.Linq
Imports DevExpress.Mvvm

Namespace ChartsDemo
    Public Class MathematicalFunctionsViewModel
        Private Const UnitFactor As Double = 15

        Private privateFunctionItemsData As List(Of Function3DItemData)
        Public Property FunctionItemsData() As List(Of Function3DItemData)
            Get
                Return privateFunctionItemsData
            End Get
            Private Set(ByVal value As List(Of Function3DItemData))
                privateFunctionItemsData = value
            End Set
        End Property
        Public Overridable Property [Function]() As Function3DItemData

        Public Sub New()
            FunctionItemsData = CreateFunctionItemsData()
            [Function] = FunctionItemsData(3)
        End Sub

        #Region "Static"

        Private Shared Function CreateFunctionItemsData() As List(Of Function3DItemData)
            Dim funcs = New Func(Of Double, Double, Point3D)() { AddressOf CalculateFirstValue, AddressOf CalculateSecondValue, AddressOf CalculateThirdValue, AddressOf CalculateFourthValue, AddressOf CalculateFifthValue }
            Dim images = Enumerable.Range(1, 5).Select(Function(x) Chart3DUtils.CreateFunctionImage(x))
            Return funcs.Zip(images, Function(func, image) New Function3DItemData With {.Image = image, .Points = CreatePoints(func)}).ToList()
        End Function

        Private Shared Function Sinc(ByVal x As Double) As Double
            Return If(x <> 0, Math.Sin(x), 1)
        End Function
        Private Shared Function CalculateFirstValue(ByVal x As Double, ByVal y As Double) As Point3D
            x *= UnitFactor
            y *= UnitFactor
            Dim value As Double = Sinc(Math.Sin(Math.Pow(Math.Pow(x, 6) + Math.Pow(y, 6), 1.0R / 6.0R))) * 5
            Return New Point3D(x, y, value)
        End Function
        Private Shared Function CalculateSecondValue(ByVal x As Double, ByVal y As Double) As Point3D
            x *= UnitFactor
            y *= UnitFactor
            Dim square As Double = Math.Sqrt(x * x + y * y)
            Dim value As Double = square * Sinc(square) * 0.2
            Return New Point3D(x, y, value)
        End Function
        Private Shared Function CalculateThirdValue(ByVal x As Double, ByVal y As Double) As Point3D
            x *= UnitFactor / 2
            y *= UnitFactor / 2
            Dim value As Double = Math.Sin(x) * Math.Cos(y) * 2
            Return New Point3D(x, y, value)
        End Function
        Private Shared Function CalculateFourthValue(ByVal x As Double, ByVal y As Double) As Point3D
            Dim theta As Double = Math.Atan2(y, x)
            Dim r As Double = x * x + y * y
            Dim z As Double = Math.Exp(-r) * Math.Sin(2 * Math.PI * Math.Sqrt(r)) * Math.Cos(3 * theta)
            Return New Point3D(x, y, z)
        End Function
        Private Shared Function CalculateFifthValue(ByVal x As Double, ByVal y As Double) As Point3D
            x *= 3
            y *= 3
            Dim z As Double = Math.Sin(x * y)
            Return New Point3D(x, y, z)
        End Function
        Private Shared Function CreatePoints(ByVal valueCalculator As Func(Of Double, Double, Point3D)) As List(Of Point3D)
            Dim points As New List(Of Point3D)()
            For x As Double = -1 To 0 Step 0.017
                For y As Double = -1 To 0 Step 0.017
                    points.Add(valueCalculator(x, y))
                Next y
            Next x
            Return points
        End Function

        #End Region
    End Class
    Public Class Function3DItemData
        Inherits BindableBase


        Private points_Renamed As List(Of Point3D)

        Private image_Renamed As BitmapImage

        Public Property Points() As List(Of Point3D)
            Get
                Return points_Renamed
            End Get
            Set(ByVal value As List(Of Point3D))
                SetProperty(points_Renamed, value, Function() Points)
            End Set
        End Property
        Public Property Image() As BitmapImage
            Get
                Return image_Renamed
            End Get
            Set(ByVal value As BitmapImage)
                SetProperty(image_Renamed, value, Function() Image)
            End Set
        End Property
    End Class
End Namespace
